{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a34194f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 14.7M  100 14.7M    0     0   644k      0  0:00:23  0:00:23 --:--:--  714k\n"
     ]
    }
   ],
   "source": [
    "#should I stay or should I go\n",
    "!rm datasets\n",
    "!curl -o datasets 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v1/linux-amd64/datasets'\n",
    "!chmod +x datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1b0958f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import papermill as pm\n",
    "import subprocess\n",
    "\n",
    "p = subprocess.Popen(\"pwd\", stdout=subprocess.PIPE, shell=True)\n",
    "pwd = p.stdout.read().decode(\"ascii\").strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6430b4a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Black is not installed, parameters wont be formatted\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e20699bdc20746fd9601a6b9fccbc430",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Executing:   0%|          | 0/16 [00:00<?, ?cell/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23/02/23 16:33:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "New version of client (14.13.1) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    65.5kB 176kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    65.5kB 176kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    164kB 326kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    197kB 389kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    328kB 521kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    426kB 669kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    459kB 706kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    590kB 894kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    655kB 867kB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    786kB 1.02MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    885kB 1.14MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.05MB 1.33MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.15MB 1.44MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.25MB 1.54MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.31MB 1.48MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.41MB 1.58MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.54MB 1.7MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.64MB 1.79MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.77MB 1.91MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    1.87MB 1.73MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.11MB 2.64MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.9MB 2.82MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    3.96MB 2.84MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.1MB 2.91MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.52MB 3.12MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    4.85MB 3.31MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.05MB 3.43MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.08MB 3.42MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.18MB 3.47MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.34MB 3.55MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.37MB 3.55MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.44MB 3.56MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.54MB 3.6MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.7MB 3.68MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.77MB 3.71MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.83MB 3.71MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.9MB 3.74MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6MB 3.69MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.09MB 3.73MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.23MB 3.79MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.26MB 3.8MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.36MB 3.83MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.39MB 3.82MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.52MB 3.87MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.68MB 3.94MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.78MB 3.98MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.85MB 4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    6.98MB 3.97MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.01MB 3.98MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.14MB 4.02MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.24MB 4.04MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.31MB 4.06MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.34MB 4.07MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.41MB 4.06MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.54MB 4.11MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.63MB 4.14MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.77MB 4.19MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.8MB 4.2MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.83MB 4.19MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.9MB 4.19MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    7.93MB 4.2MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8MB 4.21MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.13MB 4.25MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.22MB 4.27MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.26MB 4.28MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.26MB 4.28MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.39MB 4.29MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.49MB 4.32MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.62MB 4.36MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.72MB 4.39MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.75MB 4.4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.78MB 4.37MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.88MB 4.4MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    8.95MB 4.41MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.04MB 4.43MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.11MB 4.44MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.21MB 4.47MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.21MB 4.47MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.27MB 4.45MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.4MB 4.49MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.5MB 4.52MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.63MB 4.56MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.7MB 4.58MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.73MB 4.58MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.83MB 4.59MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.9MB 4.59MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    9.96MB 4.6MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.1MB 4.62MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.1MB 4.62MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.2MB 4.64MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.2MB 4.63MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.3MB 4.65MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.4MB 4.67MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.6MB 4.72MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.7MB 4.74MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.7MB 4.74MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.8MB 4.74MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.8MB 4.74MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    10.9MB 4.76MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    11MB 4.78MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    11.1MB 4.78MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    11.1MB 4.79MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    11.2MB 4.79MB/s\n",
      "\u001b[1A\u001b[2KDownloading: deinococcus_radiodurans_complete.zip    11.2MB done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "New version of client (14.13.1) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archive:  deinococcus_radiodurans_complete.zip\n",
      "  inflating: README.md               \n",
      "  inflating: ncbi_dataset/data/assembly_data_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/GCA_008329785.1/genomic.gbff  \n",
      "  inflating: ncbi_dataset/data/GCA_011604805.1/genomic.gbff  \n",
      "  inflating: ncbi_dataset/data/GCA_011604825.1/genomic.gbff  \n",
      "  inflating: ncbi_dataset/data/GCA_020546685.1/genomic.gbff  \n",
      "  inflating: ncbi_dataset/data/GCA_024138735.1/genomic.gbff  \n",
      "  inflating: ncbi_dataset/data/GCA_008329785.1/sequence_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/GCA_011604805.1/sequence_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/GCA_011604825.1/sequence_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/GCA_020546685.1/sequence_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/GCA_024138735.1/sequence_report.jsonl  \n",
      "  inflating: ncbi_dataset/data/dataset_catalog.json  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cells': [{'cell_type': 'code',\n",
       "   'execution_count': 1,\n",
       "   'id': '6aecfabf',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:07.643171',\n",
       "     'end_time': '2023-02-23T16:33:08.308343',\n",
       "     'duration': 0.665172,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:07.652597Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:07.653415Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:08.304460Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:08.305598Z'}},\n",
       "   'outputs': [],\n",
       "   'source': 'import pandas as pd\\nimport numpy as np\\nimport os\\nfrom pathlib import Path\\nimport subprocess\\nfrom pyspark.sql import SparkSession\\nfrom pyspark.sql.types import *\\nfrom Bio import SeqIO'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 2,\n",
       "   'id': 'fd49ffca',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:08.313199',\n",
       "     'end_time': '2023-02-23T16:33:13.889469',\n",
       "     'duration': 5.57627,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:08.318906Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:08.319414Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:13.886387Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:13.885017Z'}},\n",
       "   'outputs': [{'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': 'Setting default log level to \"WARN\".\\nTo adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '23/02/23 16:33:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\\n'}],\n",
       "   'source': \"spark = SparkSession.builder.appName('MastersProject').getOrCreate()\"},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 3,\n",
       "   'id': 'f9622609',\n",
       "   'metadata': {'tags': ['parameters'],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:13.895263',\n",
       "     'end_time': '2023-02-23T16:33:13.913119',\n",
       "     'duration': 0.017856,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:13.902705Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:13.903306Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:13.908006Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:13.909617Z'}},\n",
       "   'outputs': [],\n",
       "   'source': 'taxon = \\'\\\\\"Escherichia coli\\\\\"\\'\\npwd = \"/home/infosebi/Documents/MastersProject\"'},\n",
       "  {'id': 'bf98e635',\n",
       "   'cell_type': 'code',\n",
       "   'metadata': {'tags': ['injected-parameters'],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:13.918519',\n",
       "     'end_time': '2023-02-23T16:33:13.935391',\n",
       "     'duration': 0.016872,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:13.926416Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:13.927019Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:13.931307Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:13.932547Z'}},\n",
       "   'execution_count': 4,\n",
       "   'source': '# Parameters\\ntaxon = \"\\'Deinococcus radiodurans\\'\"\\npwd = \"/home/infosebi/Documents/MastersProject\"\\n',\n",
       "   'outputs': []},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 5,\n",
       "   'id': 'fdde1723',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:13.939961',\n",
       "     'end_time': '2023-02-23T16:33:16.402302',\n",
       "     'duration': 2.462341,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:13.946432Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:13.947048Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:16.398641Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:16.399687Z'}},\n",
       "   'outputs': [{'output_type': 'stream', 'name': 'stdout', 'text': '8\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': 'New version of client (14.13.1) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets\\n'}],\n",
       "   'source': 'taxon_file = taxon.lower().replace(\" \", \"_\").replace(\"\\\\\\'\", \"\") + \"_complete\"\\nsubprocess.run(f\"\"\"./datasets summary genome taxon {str(taxon)} --assembly-level complete_genome --assembly-source genbank --annotated >  {taxon_file}.json\"\"\", shell=True)\\np = subprocess.Popen(f\"\"\"jq \\'.total_count\\' {taxon_file}.json\"\"\", stdout=subprocess.PIPE, shell=True)\\nprint(p.stdout.read().decode(\\'ascii\\').strip())'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 6,\n",
       "   'id': 'bcb6ef94',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:16.407529',\n",
       "     'end_time': '2023-02-23T16:33:16.492574',\n",
       "     'duration': 0.085045,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:16.413974Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:16.414460Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:16.488925Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:16.490001Z'}},\n",
       "   'outputs': [{'output_type': 'execute_result',\n",
       "     'metadata': {},\n",
       "     'data': {'text/plain': 'CompletedProcess(args=\"cat deinococcus_radiodurans_complete.json | jq -r \\'.assemblies[].assembly | [.assembly_accession, .org.strain, .org.tax_id, .seq_length, .gc_count, .contig_n50, .annotation_metadata.stats.gene_counts.total] | @tsv\\' | head -5 > info_table.tsv\", returncode=0)'},\n",
       "     'execution_count': 6}],\n",
       "   'source': 'subprocess.run(f\"cat {taxon_file}.json | jq -r \\'.assemblies[].assembly | [.assembly_accession, .org.strain, .org.tax_id, .seq_length, .gc_count, .contig_n50, .annotation_metadata.stats.gene_counts.total] | @tsv\\' | head -5 > info_table.tsv\", shell=True) '},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 7,\n",
       "   'id': 'e3fbbd3a',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:16.497223',\n",
       "     'end_time': '2023-02-23T16:33:16.575472',\n",
       "     'duration': 0.078249,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:16.504360Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:16.504739Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:16.571805Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:16.572929Z'}},\n",
       "   'outputs': [{'output_type': 'execute_result',\n",
       "     'metadata': {},\n",
       "     'data': {'text/plain': 'CompletedProcess(args=\"cat deinococcus_radiodurans_complete.json | jq -r \\'.assemblies[].assembly| .assembly_accession\\' | head -5 > accession_list.txt\", returncode=0)'},\n",
       "     'execution_count': 7}],\n",
       "   'source': 'subprocess.run(f\"cat {taxon_file}.json | jq -r \\'.assemblies[].assembly| .assembly_accession\\' | head -5 > accession_list.txt\", shell=True)'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 8,\n",
       "   'id': 'ad12954c',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:16.581089',\n",
       "     'end_time': '2023-02-23T16:33:20.681316',\n",
       "     'duration': 4.100227,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:16.588664Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:16.589708Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:20.673370Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:20.670296Z'}},\n",
       "   'outputs': [{'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': 'Downloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.55kB 5.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.65kB 37.4kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    32.8kB 133kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    65.5kB 176kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    65.5kB 176kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    131kB 335kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    164kB 326kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    197kB 389kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    262kB 503kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    295kB 532kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    328kB 521kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    426kB 669kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    459kB 706kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    590kB 894kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    623kB 937kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    655kB 867kB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    786kB 1.02MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    885kB 1.14MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.05MB 1.33MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.15MB 1.44MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.25MB 1.54MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.28MB 1.57MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.31MB 1.48MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.41MB 1.58MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.54MB 1.7MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.64MB 1.79MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.77MB 1.91MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.84MB 1.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    1.87MB 1.73MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    2.56MB 2.3MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.08MB 2.68MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.11MB 2.64MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.44MB 2.88MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.74MB 2.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.9MB 2.82MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    3.96MB 2.84MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.1MB 2.91MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.42MB 3.12MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.52MB 3.12MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    4.85MB 3.31MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.05MB 3.43MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.08MB 3.42MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.18MB 3.47MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.34MB 3.55MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.37MB 3.55MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.44MB 3.56MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.54MB 3.6MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.7MB 3.68MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.77MB 3.71MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.83MB 3.71MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.9MB 3.74MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    5.96MB 3.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6MB 3.69MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.09MB 3.73MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.23MB 3.79MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.26MB 3.8MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.36MB 3.83MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.39MB 3.82MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.52MB 3.87MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.68MB 3.94MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.78MB 3.98MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.85MB 4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.88MB 4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    6.98MB 3.97MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.01MB 3.98MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.14MB 4.02MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.24MB 4.04MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.31MB 4.06MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.34MB 4.07MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.41MB 4.06MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.54MB 4.11MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.63MB 4.14MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.77MB 4.19MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.8MB 4.2MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.83MB 4.19MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.9MB 4.19MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    7.93MB 4.2MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8MB 4.21MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.13MB 4.25MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.22MB 4.27MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.26MB 4.28MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.26MB 4.28MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.39MB 4.29MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.49MB 4.32MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.62MB 4.36MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.72MB 4.39MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.75MB 4.4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.78MB 4.37MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.88MB 4.4MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    8.95MB 4.41MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.04MB 4.43MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.11MB 4.44MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.21MB 4.47MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.21MB 4.47MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.27MB 4.45MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.4MB 4.49MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.5MB 4.52MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.63MB 4.56MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.7MB 4.58MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.73MB 4.58MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.83MB 4.59MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.9MB 4.59MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    9.96MB 4.6MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.1MB 4.62MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.1MB 4.62MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.2MB 4.64MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.2MB 4.63MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.3MB 4.65MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.4MB 4.67MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.6MB 4.72MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.7MB 4.74MB/s\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.7MB 4.74MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.8MB 4.74MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.8MB 4.74MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    10.9MB 4.76MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    11MB 4.78MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    11.1MB 4.78MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    11.1MB 4.79MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    11.2MB 4.79MB/s\\n\\x1b[1A\\x1b[2KDownloading: deinococcus_radiodurans_complete.zip    11.2MB done\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': 'New version of client (14.13.1) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets\\n'},\n",
       "    {'output_type': 'execute_result',\n",
       "     'metadata': {},\n",
       "     'data': {'text/plain': \"CompletedProcess(args='./datasets download genome accession --exclude-genomic-cds --exclude-gff3 --exclude-protein --exclude-rna --exclude-seq --include-gbff --inputfile accession_list.txt --filename deinococcus_radiodurans_complete.zip', returncode=0)\"},\n",
       "     'execution_count': 8}],\n",
       "   'source': 'subprocess.run(f\"./datasets download genome accession --exclude-genomic-cds --exclude-gff3 --exclude-protein --exclude-rna --exclude-seq --include-gbff --inputfile accession_list.txt --filename {taxon_file}.zip\", shell=True)'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 9,\n",
       "   'id': 'c20c7e4d',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:20.701131',\n",
       "     'end_time': '2023-02-23T16:33:21.730437',\n",
       "     'duration': 1.029306,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:20.724986Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:20.725989Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:21.726652Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:21.727867Z'}},\n",
       "   'outputs': [{'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': 'Archive:  deinococcus_radiodurans_complete.zip\\n  inflating: README.md               \\n  inflating: ncbi_dataset/data/assembly_data_report.jsonl  \\n  inflating: ncbi_dataset/data/GCA_008329785.1/genomic.gbff  '},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\n  inflating: ncbi_dataset/data/GCA_011604805.1/genomic.gbff  '},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\n  inflating: ncbi_dataset/data/GCA_011604825.1/genomic.gbff  '},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\n  inflating: ncbi_dataset/data/GCA_020546685.1/genomic.gbff  '},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\n  inflating: ncbi_dataset/data/GCA_024138735.1/genomic.gbff  '},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '\\n  inflating: ncbi_dataset/data/GCA_008329785.1/sequence_report.jsonl  \\n  inflating: ncbi_dataset/data/GCA_011604805.1/sequence_report.jsonl  \\n  inflating: ncbi_dataset/data/GCA_011604825.1/sequence_report.jsonl  \\n  inflating: ncbi_dataset/data/GCA_020546685.1/sequence_report.jsonl  \\n  inflating: ncbi_dataset/data/GCA_024138735.1/sequence_report.jsonl  \\n  inflating: ncbi_dataset/data/dataset_catalog.json  \\n'},\n",
       "    {'output_type': 'execute_result',\n",
       "     'metadata': {},\n",
       "     'data': {'text/plain': \"CompletedProcess(args='unzip deinococcus_radiodurans_complete.zip', returncode=0)\"},\n",
       "     'execution_count': 9}],\n",
       "   'source': '!rm README.md\\n!rm -r ncbi_dataset\\nsubprocess.run(f\"unzip {taxon_file}.zip\", shell=True)'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 10,\n",
       "   'id': 'cf88561e',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:21.738869',\n",
       "     'end_time': '2023-02-23T16:33:21.772461',\n",
       "     'duration': 0.033592,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:21.748120Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:21.748539Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:21.767236Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:21.768771Z'}},\n",
       "   'outputs': [],\n",
       "   'source': 'df = pd.read_table(\"info_table.tsv\", header=None)\\ndf.columns = [\\'Genome\\', \\'Strain\\', \\'Taxonomy Id\\', \\'Sequence length\\', \"GC Percent (%)\", \"ContigN50\", \"Gene count\"]\\ndf.set_index(\"Genome\", inplace = True)\\ndf[\"rRNA\"]=np.nan\\ndf[\"tRNA\"]=np.nan'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 11,\n",
       "   'id': '695e6d75',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:21.783644',\n",
       "     'end_time': '2023-02-23T16:33:22.035969',\n",
       "     'duration': 0.252325,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:21.795220Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:21.795706Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:22.030684Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:22.031882Z'}},\n",
       "   'outputs': [],\n",
       "   'source': 'for i, row in df.iterrows():\\n    os.chdir(pwd + f\"/ncbi_dataset/data/{i}\")\\n    p = subprocess.Popen(\"\"\"cat genomic.gbff | grep rRNAs | head -1 | awk -F\":: \" \\'{ print $2}\\'\"\"\", stdout=subprocess.PIPE, shell=True)\\n    rRNA = p.stdout.read().decode(\\'ascii\\').strip()\\n    p = subprocess.Popen(\"\"\"cat genomic.gbff | grep tRNAs | head -1 | awk -F\":: \" \\'{ print $2}\\'\"\"\", stdout=subprocess.PIPE, shell=True)\\n    tRNA = p.stdout.read().decode(\\'ascii\\').strip()\\n    df.at[i, \"rRNA\"] = rRNA\\n    df.at[i, \"tRNA\"] = tRNA\\nos.chdir(pwd)'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 12,\n",
       "   'id': 'f55f518c',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:22.047290',\n",
       "     'end_time': '2023-02-23T16:33:22.105782',\n",
       "     'duration': 0.058492,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:22.062123Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:22.063826Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:22.098902Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:22.095854Z'}},\n",
       "   'outputs': [{'output_type': 'execute_result',\n",
       "     'metadata': {},\n",
       "     'data': {'text/plain': '                     Strain  Taxonomy Id  Sequence length  GC Percent (%)  \\\\\\nGenome                                                                      \\nGCA_008329785.1      R1 dM1         1299          3283189         2188402   \\nGCA_011604805.1      BNK-50         1299          3284329         2188764   \\nGCA_011604825.1      BND-54         1299          3283098         2188024   \\nGCA_024138735.1      BRD125         1299          3281648         2186593   \\nGCA_020546685.1  ATCC 13939      1408434          3279598         2186449   \\n\\n                 ContigN50  Gene count                    rRNA tRNA  \\nGenome                                                               \\nGCA_008329785.1    2647538        3309  3, 3, 3 (5S, 16S, 23S)   50  \\nGCA_011604805.1    2648775        3202  3, 3, 3 (5S, 16S, 23S)   50  \\nGCA_011604825.1    2648637        3209  3, 3, 3 (5S, 16S, 23S)   50  \\nGCA_024138735.1    2647192        3543  3, 3, 3 (5S, 16S, 23S)   49  \\nGCA_020546685.1    2644543        3305  3, 3, 3 (5S, 16S, 23S)   50  ',\n",
       "      'text/html': '<div>\\n<style scoped>\\n    .dataframe tbody tr th:only-of-type {\\n        vertical-align: middle;\\n    }\\n\\n    .dataframe tbody tr th {\\n        vertical-align: top;\\n    }\\n\\n    .dataframe thead th {\\n        text-align: right;\\n    }\\n</style>\\n<table border=\"1\" class=\"dataframe\">\\n  <thead>\\n    <tr style=\"text-align: right;\">\\n      <th></th>\\n      <th>Strain</th>\\n      <th>Taxonomy Id</th>\\n      <th>Sequence length</th>\\n      <th>GC Percent (%)</th>\\n      <th>ContigN50</th>\\n      <th>Gene count</th>\\n      <th>rRNA</th>\\n      <th>tRNA</th>\\n    </tr>\\n    <tr>\\n      <th>Genome</th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n      <th></th>\\n    </tr>\\n  </thead>\\n  <tbody>\\n    <tr>\\n      <th>GCA_008329785.1</th>\\n      <td>R1 dM1</td>\\n      <td>1299</td>\\n      <td>3283189</td>\\n      <td>2188402</td>\\n      <td>2647538</td>\\n      <td>3309</td>\\n      <td>3, 3, 3 (5S, 16S, 23S)</td>\\n      <td>50</td>\\n    </tr>\\n    <tr>\\n      <th>GCA_011604805.1</th>\\n      <td>BNK-50</td>\\n      <td>1299</td>\\n      <td>3284329</td>\\n      <td>2188764</td>\\n      <td>2648775</td>\\n      <td>3202</td>\\n      <td>3, 3, 3 (5S, 16S, 23S)</td>\\n      <td>50</td>\\n    </tr>\\n    <tr>\\n      <th>GCA_011604825.1</th>\\n      <td>BND-54</td>\\n      <td>1299</td>\\n      <td>3283098</td>\\n      <td>2188024</td>\\n      <td>2648637</td>\\n      <td>3209</td>\\n      <td>3, 3, 3 (5S, 16S, 23S)</td>\\n      <td>50</td>\\n    </tr>\\n    <tr>\\n      <th>GCA_024138735.1</th>\\n      <td>BRD125</td>\\n      <td>1299</td>\\n      <td>3281648</td>\\n      <td>2186593</td>\\n      <td>2647192</td>\\n      <td>3543</td>\\n      <td>3, 3, 3 (5S, 16S, 23S)</td>\\n      <td>49</td>\\n    </tr>\\n    <tr>\\n      <th>GCA_020546685.1</th>\\n      <td>ATCC 13939</td>\\n      <td>1408434</td>\\n      <td>3279598</td>\\n      <td>2186449</td>\\n      <td>2644543</td>\\n      <td>3305</td>\\n      <td>3, 3, 3 (5S, 16S, 23S)</td>\\n      <td>50</td>\\n    </tr>\\n  </tbody>\\n</table>\\n</div>'},\n",
       "     'execution_count': 12}],\n",
       "   'source': 'df'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 13,\n",
       "   'id': '05f03976',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:22.124381',\n",
       "     'end_time': '2023-02-23T16:33:31.477096',\n",
       "     'duration': 9.352715,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:22.145256Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:22.146039Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:31.471169Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:31.472719Z'}},\n",
       "   'outputs': [{'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': '/home/infosebi/.local/lib/python3.10/site-packages/pyspark/sql/pandas/conversion.py:474: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.\\n  for column, series in pdf.iteritems():\\n/home/infosebi/.local/lib/python3.10/site-packages/pyspark/sql/pandas/conversion.py:486: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.\\n  for column, series in pdf.iteritems():\\n'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': '\\r[Stage 0:>                                                          (0 + 1) / 1]\\r'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stderr',\n",
       "     'text': '\\r                                                                                \\r'},\n",
       "    {'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '+---------------+----------+-----------+---------------+--------------+---------+----------+--------------------+----+\\n|         Genome|    Strain|Taxonomy Id|Sequence length|GC Percent (%)|ContigN50|Gene count|                rRNA|tRNA|\\n+---------------+----------+-----------+---------------+--------------+---------+----------+--------------------+----+\\n|GCA_008329785.1|    R1 dM1|       1299|        3283189|       2188402|  2647538|      3309|3, 3, 3 (5S, 16S,...|  50|\\n|GCA_011604805.1|    BNK-50|       1299|        3284329|       2188764|  2648775|      3202|3, 3, 3 (5S, 16S,...|  50|\\n|GCA_011604825.1|    BND-54|       1299|        3283098|       2188024|  2648637|      3209|3, 3, 3 (5S, 16S,...|  50|\\n|GCA_024138735.1|    BRD125|       1299|        3281648|       2186593|  2647192|      3543|3, 3, 3 (5S, 16S,...|  49|\\n|GCA_020546685.1|ATCC 13939|    1408434|        3279598|       2186449|  2644543|      3305|3, 3, 3 (5S, 16S,...|  50|\\n+---------------+----------+-----------+---------------+--------------+---------+----------+--------------------+----+\\n\\n'}],\n",
       "   'source': 'df.reset_index(drop=False,inplace=True)\\nsparkDF_info=spark.createDataFrame(df) \\nsparkDF_info.show()'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 14,\n",
       "   'id': '2239508b',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:31.486544',\n",
       "     'end_time': '2023-02-23T16:33:36.912190',\n",
       "     'duration': 5.425646,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:31.496843Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:31.497247Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:36.907819Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:36.909143Z'}},\n",
       "   'outputs': [{'output_type': 'stream',\n",
       "     'name': 'stdout',\n",
       "     'text': '+----+-------+---------+--------------------+----------+\\n|Name| Length|Start Pos|          Annotation|Complement|\\n+----+-------+---------+--------------------+----------+\\n|   -|2647538|        0|branched-chain am...|      true|\\n|cysK|    953|      472| cysteine synthase A|     false|\\n|   -|   1680|     1614|glycerol-3-phosph...|     false|\\n|   -|    918|     3415|alpha/beta fold h...|     false|\\n|   -|   1395|     4402|MATE family efflu...|      true|\\n|   -|    549|     5816|dihydrofolate red...|      true|\\n|   -|   1398|     6453|aspartate aminotr...|      true|\\n|   -|    636|     7835|hypothetical protein|      true|\\n|   -|    561|     8474| N-acetyltransferase|      true|\\n|   -|    477|     9090| N-acetyltransferase|      true|\\n|   -|    624|     9610| N-acetyltransferase|      true|\\n|   -|   1197|    10241|gfo/Idh/MocA fami...|     false|\\n|   -|    183|    11696|hypothetical protein|      true|\\n|   -|   1233|    11928|lycopene cyclase ...|      true|\\n|   -|    741|    13246|   anti-sigma factor|      true|\\n|   -|    849|    13983|sigma-70 family R...|      true|\\n|tatA|    399|    14999|twin-arginine tra...|     false|\\n|tatC|    813|    15438|twin-arginine tra...|     false|\\n| lgt|   1020|    16296|prolipoprotein di...|     false|\\n|   -|   1464|    17445|bifunctional N-ac...|     false|\\n+----+-------+---------+--------------------+----------+\\nonly showing top 20 rows\\n\\n'},\n",
       "    {'output_type': 'stream', 'name': 'stdout', 'text': '(16259, 5)\\n'}],\n",
       "   'source': '#gene length, gene start pos, annotation(product), direction(complement)\\n\\ncolumns = StructType([StructField(\\'Name\\', StringType(), False),\\n                       StructField(\\'Length\\', IntegerType(), False),\\n                       StructField(\\'Start Pos\\', IntegerType(), False),\\n                       StructField(\\'Annotation\\', StringType(), False),\\n                       StructField(\\'Complement\\', BooleanType(), False),])\\n\\n\\nadded_row=[]\\nfor row_iterator in sparkDF_info.collect():\\n    file_path = pwd + f\"/ncbi_dataset/data/{row_iterator[\\'Genome\\']}/genomic.gbff\"\\n    for gb_obj in SeqIO.parse(file_path,\\'genbank\\'):\\n        all_cds = []\\n        for feature in gb_obj.features:\\n            if feature.type == \"CDS\":\\n                all_cds.append(feature)\\n        for cds in all_cds:\\n            if \\'gene\\' in cds.qualifiers:\\n                gene = cds.qualifiers[\\'gene\\'][0]\\n            else:\\n                gene = \"-\"\\n            complement = False\\n            if cds.location.strand < 1:\\n                complement = True\\n            added_row.append([gene, cds.location.end-cds.location.start, int(cds.location.start) ,cds.qualifiers[\\'product\\'][0], complement])\\nsparkDF_cds = spark.createDataFrame(added_row, columns)\\nsparkDF_cds.show()\\nprint((sparkDF_cds.count(), len(sparkDF_cds.columns)))\\n#forward, reverse\\n#ec number\\n#java 19\\n#java jar cluster zip'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 15,\n",
       "   'id': '1d7adcb4',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:36.927364',\n",
       "     'end_time': '2023-02-23T16:33:36.954455',\n",
       "     'duration': 0.027091,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:36.943988Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:36.944860Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:36.949014Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:36.950261Z'}},\n",
       "   'outputs': [],\n",
       "   'source': '#genbank file parser, GBParsyPy, BioPython\\n#is it possible to have interaction in jupyter notebook\\n#install Eggnog database -> run Eggnog annotation on gbff file.\\n#google collab (vs-code, github repo)\\n#circus (input-data?)\\n#sigmarkarl github user'},\n",
       "  {'cell_type': 'code',\n",
       "   'execution_count': 16,\n",
       "   'id': '4e4b43ef',\n",
       "   'metadata': {'tags': [],\n",
       "    'papermill': {'exception': False,\n",
       "     'start_time': '2023-02-23T16:33:36.962512',\n",
       "     'end_time': '2023-02-23T16:33:36.978545',\n",
       "     'duration': 0.016033,\n",
       "     'status': 'completed'},\n",
       "    'execution': {'iopub.status.busy': '2023-02-23T16:33:36.972062Z',\n",
       "     'iopub.execute_input': '2023-02-23T16:33:36.972423Z',\n",
       "     'shell.execute_reply': '2023-02-23T16:33:36.975075Z',\n",
       "     'iopub.status.idle': '2023-02-23T16:33:36.975869Z'}},\n",
       "   'outputs': [],\n",
       "   'source': '#criteria for genome selection to strict (assembly-source genbank)?\\n#take care of warnings\\n#GenomeDiagramm library'}],\n",
       " 'metadata': {'celltoolbar': 'Tags',\n",
       "  'kernelspec': {'display_name': 'Python 3 (ipykernel)',\n",
       "   'language': 'python',\n",
       "   'name': 'python3'},\n",
       "  'language_info': {'name': 'python',\n",
       "   'version': '3.10.9',\n",
       "   'mimetype': 'text/x-python',\n",
       "   'codemirror_mode': {'name': 'ipython', 'version': 3},\n",
       "   'pygments_lexer': 'ipython3',\n",
       "   'nbconvert_exporter': 'python',\n",
       "   'file_extension': '.py'},\n",
       "  'papermill': {'default_parameters': {},\n",
       "   'parameters': {'taxon': \"'Deinococcus radiodurans'\",\n",
       "    'pwd': '/home/infosebi/Documents/MastersProject'},\n",
       "   'environment_variables': {},\n",
       "   'version': '2.4.0',\n",
       "   'input_path': '/home/infosebi/Documents/MastersProject/MastersProject.ipynb',\n",
       "   'output_path': '/home/infosebi/Documents/MastersProject/ParameterizedMastersProject.ipynb',\n",
       "   'start_time': '2023-02-23T16:33:06.447054',\n",
       "   'end_time': '2023-02-23T16:33:39.611428',\n",
       "   'duration': 33.164374,\n",
       "   'exception': None}},\n",
       " 'nbformat': 4,\n",
       " 'nbformat_minor': 5}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pm.execute_notebook(\n",
    "   pwd+'/MastersProject.ipynb',\n",
    "   pwd+'/ParameterizedMastersProject.ipynb',\n",
    "   parameters=dict(taxon=\"\\'Deinococcus radiodurans\\'\", pwd=pwd)\n",
    ")\n",
    "#draft sequence\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4746c80c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
